File: Java/merge_sort.java
public class merge_sort {
    public static void mergeSort(int arr[], int si, int ei){
        printArr(arr, si, ei);
        if(si>=ei){
            return;
        }
        int mid = si + (ei-si)/2;
        mergeSort(arr, si, mid);
        mergeSort(arr, mid+1, ei);
        merge(arr, si, mid, ei);
        printArr(arr, 0, 5);
    }
    public static void merge(int arr[], int si, int mid, int ei){
        int temp[] = new int[ei-si+1];
        int i = si;
        int j = mid+1;
        int k = 0;
        while(i <= mid && j <= ei){
            if(arr[i] < arr[j]){
                temp[k++] = arr[i++];
            } else {
                temp[k++] = arr[j++];
            }
        }
        while(i <= mid){
            temp[k++] = arr[i++];
        }
        while(j <= ei){
            temp[k++] = arr[j++];
        }
        for(int oi=si, ti=0; ti<temp.length; oi++, ti++){
            arr[oi] = temp[ti];
        }
    }
    public static void printArr(int arr[], int si, int ei){
        for(int i=si; i<=ei; i++){
            if(i==ei){
                System.out.println(arr[i]);
            } else {
                System.out.print(arr[i] + ", ");
            }
        }
    }
    public static void main(String args[]){
        int sample[] = {6, 5, 4, 3, 2, 1};
        mergeSort(sample, 0, sample.length-1);
        printArr(sample, 0, sample.length-1);
    }
}

File: CPP/mergeSort.cpp
#include <iostream>
using namespace std;

void mergeSort(int[], int, int);

int main() {
    int arr[6] = {6, 2, 3 ,5, 4, 1};
    mergeSort(arr, 0, 5);

    for(int i = 0; i < 6; i++) {
        cout << arr[i] << ", ";
    }
    cout << endl;
}

void mergeSort(int arr[], int start, int end) {
    if(start >= end) {
        return;
    }
    int mid = start + (end-start)/2;
    mergeSort(arr, start, mid);
    mergeSort(arr, mid+1, end);
    
    int len = end-start+1;
    int ind = 0, i = start, j = mid+1;
    int temp[len];
    while(i <= mid && j <= end) {
        if(arr[i] <= arr[j]) {
            temp[ind++] = arr[i++];
        }
        else{
            temp[ind++] = arr[j++];
        }
    }
    while(i <= mid) {
        temp[ind++] = arr[i++];
    }
    while(j <= end) {
        temp[ind++] = arr[j++];
    }
    for(int i = 0, j = start; i < len; i++, j++) {
        arr[j] = temp[i];
    }
}